File: ws/diaz/u3/ws24MyFirstFrontEnd/ws24MyFirstFrontEnd/crud/src/routes/supplyRoutes.ts
import { Router, Request, Response } from 'express';
import { Prisma } from '@prisma/client';
import prisma from '../db';

const supplyRouter = Router();

supplyRouter.get('/supplies', async (req: Request, res: Response) => {
  try {
    const inventory = await prisma.supplies.findMany();
    return res.status(200).json(inventory);
  } catch {
    return res.status(500).json({ error: "Unable to fetch inventory." });
  }
});

supplyRouter.get('/supplies/quantity-thresholds/:maxQuantity', async (req: Request, res: Response) => {
  try {
    const maxQuantity = parseInt(String(req.params.maxQuantity), 10);
    if (isNaN(maxQuantity)) {
      return res.status(400).json({ error: "Invalid maximum quantity format." });
    }
    const filteredSupplies = await prisma.supplies.findMany({
      where: {
        quantity: {
          lte: maxQuantity,
        },
      },
    });
    return res.status(200).json(filteredSupplies);
  } catch {
    return res.status(500).json({ error: "Database error while filtering supply quantities." });
  }
});

supplyRouter.get('/supplies/statuses/:statusValue', async (req: Request, res: Response) => {
  try {
    const statusValue = String(req.params.statusValue);
    if (statusValue !== 'Expired' && statusValue !== 'NextExpiration' && statusValue !== 'Current') {
      return res.status(400).json({ error: "Provided status value does not exist." });
    }
    const filteredSupplies = await prisma.supplies.findMany({
      where: {
        status: statusValue,
      },
    });
    return res.status(200).json(filteredSupplies);
  } catch {
    return res.status(500).json({ error: "Database error while filtering supply statuses." });
  }
});

supplyRouter.post('/supply', async (req: Request, res: Response) => {
  try {
    const { supplyName, quantity, unitCost, orderDate, expirationDate } = req.body;
    if (!supplyName || quantity === undefined || unitCost === undefined || !orderDate || !expirationDate) {
      return res.status(400).json({ error: "Missing supply data." });
    }
    await prisma.supplies.create({
      data: {
        supplyName,
        quantity: parseInt(quantity, 10),
        unitCost: new Prisma.Decimal(unitCost),
        orderDate: new Date(orderDate),
        expirationDate: new Date(expirationDate),
        status: 'Current',
      },
    });
    return res.status(201).json({ success: true, message: "Supply added" });
  } catch {
    return res.status(500).json({ error: "Could not add supply." });
  }
});

supplyRouter.put('/supplies/:id', async (req: Request, res: Response) => {
  try {
    const id = req.params.id;
    const { supplyName, quantity, unitCost, orderDate, expirationDate } = req.body;
    let supplyId: bigint;
    try {
      supplyId = BigInt(String(id));
    } catch {
      return res.status(400).json({ error: "Invalid supply ID or quantity." });
    }
    if (quantity !== undefined && (isNaN(Number(quantity)) || quantity < 0)) {
      return res.status(400).json({ error: "Invalid supply ID or quantity." });
    }
    const updateData: any = {};
    if (supplyName !== undefined) updateData.supplyName = supplyName;
    if (quantity !== undefined) updateData.quantity = parseInt(quantity, 10);
    if (unitCost !== undefined) updateData.unitCost = new Prisma.Decimal(unitCost);
    if (orderDate !== undefined) updateData.orderDate = new Date(orderDate);
    if (expirationDate !== undefined) updateData.expirationDate = new Date(expirationDate);

    await prisma.supplies.update({
      where: { id: supplyId },
      data: updateData,
    });
    return res.status(200).json({ success: true, message: "Supply updated" });
  } catch (error) {
    if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === 'P2025') {
      return res.status(404).json({ error: "Supply not found." });
    }
    return res.status(400).json({ error: "Invalid supply ID or quantity." });
  }
});

supplyRouter.delete('/supplies/:id', async (req: Request, res: Response) => {
  try {
    const id = req.params.id;
    let supplyId: bigint;
    try {
      supplyId = BigInt(String(id));
    } catch {
      return res.status(400).json({ error: "Supply ID required." });
    }
    await prisma.supplies.delete({
      where: { id: supplyId },
    });
    return res.status(200).json({ success: true, message: "Supply deleted" });
  } catch (error) {
    if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === 'P2025') {
      return res.status(404).json({ error: "Supply not found." });
    }
    return res.status(500).json({ error: "Deletion failed." });
  }
});

export default supplyRouter;

File: hw/cardenas/u3/hw21PythonAPI/routes/supply_routes.py
from datetime import datetime
from typing import Literal
from decimal import Decimal
from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse
from sqlmodel import Session, select

from database import get_db
from models import Supply, SupplyRequest 

router = APIRouter()

def clean_date(raw_date):
    text_date = str(raw_date)
    if "T" in text_date:
        text_date = text_date.split("T")[0]
    return text_date


@router.get("/supplies")
def get_supplies(db: Session = Depends(get_db)):
    try:
        inventory = db.exec(select(Supply)).all()
        return inventory
    except Exception:
        return JSONResponse(status_code=500, content={"error": "Unable to fetch inventory."})


@router.get("/supplies/quantity-thresholds/{maxQuantity}")
def get_supplies_quantity_thresholds(maxQuantity: int, db: Session = Depends(get_db)):
    try:
        inventory = db.exec(select(Supply).where(Supply.quantity <= maxQuantity)).all()
        return inventory
    except Exception:
        return JSONResponse(status_code=500, content={"error": "Database error while filtering supply quantities."})


@router.get("/supplies/statuses/{statusValue}")
def get_supplies_statuses(statusValue: Literal["Expired", "NextExpiration", "Current"], db: Session = Depends(get_db)):
    try:
        inventory = db.exec(select(Supply).where(Supply.status == statusValue)).all()
        return inventory
    except Exception:
        return JSONResponse(status_code=500, content={"error": "Database error while filtering supply statuses."})


@router.post("/supply")
def create_supply(data: SupplyRequest, db: Session = Depends(get_db)):
    if not data.supplyName or data.quantity is None or data.unitCost is None or not data.orderDate or not data.expirationDate:
        return JSONResponse(status_code=400, content={"error": "Missing supply data."})

    try:
        new_supply = Supply(
            supplyName=data.supplyName,
            quantity=data.quantity,
            unitCost=data.unitCost,
            orderDate=clean_date(data.orderDate),
            expirationDate=clean_date(data.expirationDate),
            status="Current"
        )
        db.add(new_supply)
        db.commit()
        return JSONResponse(status_code=201, content={"success": True, "message": "Supply added"})
    except Exception:
        return JSONResponse(status_code=400, content={"error": "Missing supply data."})


@router.put("/supplies/{id}")
def update_supply(id: int, data: SupplyRequest, db: Session = Depends(get_db)):
    db_supply = db.get(Supply, id)
    if not db_supply:
        return JSONResponse(status_code=404, content={"error": "Supply not found."})

    if not data.supplyName or data.quantity is None or data.unitCost is None or not data.orderDate or not data.expirationDate:
        return JSONResponse(status_code=400, content={"error": "Invalid supply ID or quantity."})

    if data.quantity < 0:
        return JSONResponse(status_code=400, content={"error": "Invalid supply ID or quantity."})

    try:
        db_supply.supplyName = data.supplyName
        db_supply.quantity = data.quantity
        db_supply.unitCost = Decimal(str(data.unitCost))
        db_supply.orderDate = clean_date(data.orderDate)
        db_supply.expirationDate = clean_date(data.expirationDate)
        db_supply.updated_at = datetime.now()

        db.add(db_supply)
        db.commit()
        return JSONResponse(status_code=200, content={"success": True, "message": "Supply updated"})
    except Exception:
        return JSONResponse(status_code=400, content={"error": "Invalid supply ID or quantity."})


@router.delete("/supplies/{id}")
def delete_supply(id: int, db: Session = Depends(get_db)):
    db_supply = db.get(Supply, id)
    if not db_supply:
        return JSONResponse(status_code=404, content={"error": "Supply not found."})

    try:
        db.delete(db_supply)
        db.commit()
        return JSONResponse(status_code=200, content={"success": True, "message": "Supply deleted"})
    except Exception:
        return JSONResponse(status_code=500, content={"error": "Deletion failed."})